### PR TITLE
[RFC] Parameterize context type of decorators

### PR DESCRIPTION
**Pull request by jodersky**
This allows customizing the way by-type parameters can be read: instead of always reading data from a `cask.Request`, this allows a decorator to parameterize the context type and pass it in explicitly from the `wrapFunction` method.

Essentially, this makes the context parameter of `ArgReader`s (which are used to translate data from a HTTP request to a scala parameter) customizable for every decorator.

## Motivation

The idea behind this proposal is to uniformize the way "named" and "by-type" parameters within the same endpoint are handled. By "named" parameters I mean parameters which are set from the `Map[String, Input]` in the delegate function, and by `by-type` parameters I mean parameters which are computed from an arity zero `ArgReader`, and thus use the `cask.Request` context to compute their value. 

E.g. 

```
@cask.get(/:foo/:bar)
def index(foo: String, bar: Int, cookie1: cask.Cookies, req: cask.Request)
```

In this case, `foo` and `bar`  and "named" and `cookie1` and `req` are "by-type".

Right now, the `wrapFunction` method handles how named parameters are set, and therefore can centrally do arbitrary pre- and post-processing. However, by-type parameters must always use a `cask.Request` to compute any values, which can be problematic if the computation is expensive or if any kind of state is maintained.

**This pull request removes this asymmetry, by allowing the `wrapFunction` to compute a custom context which is computed once before being passed to all `InputReader`s**

## Implementation approach

- add `InputContext` type parameter to decorators
- add a parameter of type `InputContext` to `delegate`. `type Delegate = (InputContext, Map[String, Input)) => ...`
- change `Decorator.invoke()` to carry a list of input contexts, one for each parameter list
- also change the `Entrypoint` macros to use these input contexts rather than hardcoding `cask.Request` (currently I've only done the work to support Scala 3 for this)

 ---

This is a proof of concept PR. As of this writing, the code only works for Scala 3.

### CODE REVIEW

File: cask/src/cask/router/Decorators.scala [hunk 1]
```diff
@@ -14,10 +15,10 @@ import cask.model.{Request, Response}
   * to `wrapFunction`, which takes a `Map` representing any additional argument
   * lists (if any).
   */
-trait Decorator[OuterReturned, InnerReturned, Input] extends scala.annotation.Annotation {
+trait Decorator[OuterReturned, InnerReturned, Input, InputContext] extends scala.annotation.Annotation {
```
Comment (jodersky) [inline cask/src/cask/router/Decorators.scala:18]: this is the main change here, along with adding an extra parameter to `delegate` and changing invocation to accommodate these changes

File: example/decoratedContext/app/src/DecoratedContext.scala [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+  @custom()
+  @cask.get("/hello/:world")
+  def hello(world: String, req: cask.Request)(session: Session, user: Int) = {
+    session.data("hello") = "world"
+    world + user
+  }
+
+  initialize()
+}
```
Comment (jodersky) [inline example/decoratedContext/app/src/DecoratedContext.scala:55]: See this as a motivating example

**Comment by lihaoyi** [comment]
I think this looks reasonable

**Comment by jodersky** [comment]
@lihaoyi, I ported the changes to also work with scala 2, and included the example in the published ones. Let me know if you think this is now good to merge.

**Review by lihaoyi — APPROVED** [review]

**Comment by lihaoyi** [comment]
@jodersky looks good, you have commit access so go ahead and merge this and the other PRs when you're happy and tag a release

**Pull request merged**